File: Java/Decorator/CoffeeShop/src/com/ericgtkb/Beverage.java
package com.ericgtkb;

// Abstract component class
public abstract class Beverage {
    String description = "Unknown Beverage";

    public String getDescription() {
        return description;
    }

    public abstract double cost();

    public void beverageSummary() {
        System.out.format("This is a(n) %s. The total cost is %.2f.\n", getDescription(), cost());
    }
}

// Concrete Beverage classes
class Espresso extends Beverage {
    public Espresso() {
        description = "Espresso";
    }

    @Override
    public double cost() {
        return 2.25;
    }
}

class HouseBlend extends Beverage {
    public HouseBlend() {
        description = "House Blend Coffee";
    }

    @Override
    public double cost() {
        return 2.93;
    }
}

File: C++/Decorator/CoffeeShop/beverage.cc
#include <iostream>
#include <iomanip>
#include "beverage.h"

std::string Beverage::get_description() const {
    return description;
}

void Beverage::beverage_summary() {
    std::cout << "This is a(n) " << get_description() << ". The total cost is "
              << std::fixed << std::setprecision(2) << cost() << ".\n";

}

Espresso::Espresso() {
    description = "Espresso";
}

double Espresso::cost() {
    return 2.25;
}

HouseBlend::HouseBlend() {
    description = "House Blend Coffee";
}

double HouseBlend::cost() {
    return 2.93;
}
